lib/id3c/cli/command/geocode.py

```diff
@@ -0,0 +1,435 @@
+          secondary: "ADD_LINE_2"
+          city: "CITY"
+          state: "ABBR"
+          zipcode: "ZIP"
+        ...
+
+    Relative paths in <config.yaml> are treated relative to the containing
+    directory of the configuration file itself.
+
+    Requires two environment variables: SMARTYSTREETS_AUTH_ID and
+    SMARTYSTREETS_AUTH_TOKEN.
+
+    Geocoded addresses are output to stdout as comma-separated values, with
+    extra values for addresses latitude, longitude, and canonicalized address.
+    If an address is invalid, then these extra columns are left blank.
+    """
+    configs = list(yaml.safe_load_all(config_file))
+
+    if config_file.name != "<stdin>":
+        config_dir = dirname(config_file.name)
+
+        # dirname is the empty string if we're in the same directory as the
+        # config file.
+        if config_dir:
+            chdir(config_dir)
+
+    for config in configs:
+        try:
+            kwargs = {
+                "filename":          config["filename"],
+                "street_column":     config["columns"]["street"],
+                "city_column":       config["columns"]["city"],
+                "state_column":      config["columns"]["state"],
+                "zipcode_column":    config["columns"]["zipcode"],
+                "secondary_column":  config["columns"].get("secondary")
+            }
+        except KeyError as key:
+            LOG.error(f"Required key «{key}» missing from config {config}")
+            raise key from None
+
+        geocoded_addresses = get_geocoded_addresses(**kwargs)
+        geocoded_addresses.to_csv(sys.stdout, index=False)
+
+
+def get_geocoded_addresses(*,
+                           filename,
+                           street_column,
+                           city_column,
+                           state_column,
+                           zipcode_column,
+                           secondary_column = None):
+    """
+    Internal function powering :func: `geocode_using_options` and
+    :func: `geocode_using_config`.
+    """
+    LOG.debug(f"Reading addresses file {filename}")
+
+    address_file_df = load_file_as_dataframe(filename)
+
+    address_column_map = {
+        'street': street_column,
+        'city': city_column,
+        'state': state_column,
+        'zipcode': zipcode_column,
+        'secondary': secondary_column
+    }
+
+    addresses_df = create_standardized_address_df(address_file_df, address_column_map)
+    standardized_addresses = convert_df_to_dict_series(addresses_df)
+
+    cache = load_or_create_cache()
+
+    smartystreets_responses = pd.DataFrame(columns = ['lat','lng','canonicalized_address'])
+
+    for address in standardized_addresses:
+        response = get_response_from_cache_or_geocoding(address, cache)
+
+        smartystreets_responses = smartystreets_responses \
+            .append(response,ignore_index = True)
+
+    geocoded_addresses = address_file_df.join(smartystreets_responses)
```

Comment:
I wonder if there's a better way to geocode addresses than to convert from DataFrame -> dict, create an empty DataFrame, loop over the dict while appending to the new DataFrame, then joining to a DataFrame again. 

Perhaps we could standardize the address on the original DataFrame and use an `apply` function on it that creates several new columns from geocoding.